### PR TITLE
Integrate mysql2 driver + add local mysql server to github test action

### PR DESCRIPTION
**Pull request by seanim0920**
## Problems
* Tests ran slowly on local and CI environments
* Tests from local and CI environments would often interrupt each other due to modifying the same database

## Proposed Solution
* Use local mysql instance for local and CI environments

## Steps
* Added "mysql2" package to TiFBackendUtils
* Integrated mysql2 package with SQLExecutable
* Updated .env vars for APILambda

## Results
* Tests now work with a locally hosted mysql instance.
* To use, start local mysql instance then add the DATABASE_HOST, DATABASE_NAME, DATABASE_USERNAME and DATABASE_PASSWORD to .env in APILambda, GeocodingLambda and TiFBackendUtils.

## Other changes
* Tests now work with a locally hosted mysql instance.
* To use, add the DATABASE_HOST, DATABASE_NAME, DATABASE_USERNAME and DATABASE_PASSWORD to .env in APILambda, GeocodingLambda and TiFBackendUtils.
* Fix check constraint error message check
* Fix getAttendees flaky tests by adding delays before adding new attendees to the test event
* Replace CREATE VIEW IF NOT EXISTS with CREATE VIEW. Former operation isn't compatible with mysql 8.0
* Add log in test after creating sql tables to catch errors
* TiFBackendUtils - Move entities.ts outside of planetscale folder
* Need to close db connection after generating entities.ts
* Recreate database instead of deleting from each table before each test for simplicity & speed
* Moved create table sql from setuphooks.ts to database.ts

## Follow Ups
* Add retry if db connection fails

TASK_eoNhf72K

### CODE REVIEW

**Comment by tifapp** [comment]
Linked Ticket: https://trello.com/c/eoNhf72K

File: APILambda/events/arrivals/getUpcomingEvents.ts [hunk 1]
```diff
@@ -1,42 +1,42 @@
...
         // eslint-disable-next-line @typescript-eslint/ban-ts-comment
         // @ts-ignore
-        hasArrived: event.hasArrived === 1,
+        hasArrived: hasArrived === 1 || hasArrived === true, // hasArrived should be treated as int, but is treated as bigint if it occurs in a transaction after TiFEventView
         arrivalRadiusMeters: 500 // TODO: Parameterize
       }
     }
 
-    eventRegions[key].eventIds.push(event.id)
+    eventRegions[key].eventIds.push(id)
```
Comment (mhayes853) [inline APILambda/events/arrivals/getUpcomingEvents.ts]: A `Map` is typically better for this kind of work.

File: APILambda/test/database.ts [hunk 2]
```diff
@@ -3,16 +3,16 @@ import { fail } from "assert"
...
+    conn.executeResult("DELETE FROM pushTokens"),
+    conn.executeResult("DELETE FROM userRelations"),
+    conn.executeResult("DELETE FROM userSettings"),
+    conn.executeResult("DELETE FROM userArrivals"),
+    conn.executeResult("DELETE FROM location")
   ])
-  await conn.queryResults("DELETE FROM event")
-  await conn.queryResults("DELETE FROM user")
+  await conn.executeResult("DELETE FROM event")
+  await conn.executeResult("DELETE FROM user")
```
Comment (mhayes853) [inline APILambda/test/database.ts]: You may be able to use `DROP DATABASE` here instead. Or, you could look into a procedure that automatically deletes all records from each table without specifying any table names explicitly. That way, you don't forget to add a new table to this function.

File: APILambda/test/userFlows/events.ts [hunk 3]
```diff
@@ -34,12 +35,13 @@ export const createEventFlow = async (
...
     attendeesList.push(attendee)
 
-    await Promise.all(
-      eventIds.map((eventId) => callJoinEvent(attendee.token, eventId))
-    )
+    for (const eventId of eventIds) {
+      // delay needed for test attendees to be properly sorted by join time
+      await delay(1000)
+      await callJoinEvent(attendee.token, eventId)
+    }
```
Comment (mhayes853) [inline APILambda/test/userFlows/events.ts:44]: We could load the DB with some preset users at the start of this test by specifying their join time manually. Though I get this is supposed to be as "black-box" as possible. If you want to keep the black box aspect, then you could lower the delay to 10ms.

File: TiFBackendUtils/SQLExecutable/utils.ts [hunk 4]
```diff
@@ -10,19 +10,47 @@ import { AwaitableResult, failure, promiseResult, success } from "../result.js"
...
   insertId: string
   rowsAffected: number
 }
 
+const typecasts: Record<number, (value: string | null) => unknown> = {
+  3: (value) => parseInt(value ?? "0") > 0, // INT or LONG as BOOLEAN
+  1: (value) => parseInt(value ?? "0") > 0, // TINYINT as BOOLEAN
+  12: (value) => value ? new Date(value) : value, // DATETIME
+  246: (value) => value ? parseFloat(value) : value // NEWDECIMAL (DECIMAL/NUMERIC)
+}
```
Comment (mhayes853) [inline TiFBackendUtils/SQLExecutable/utils.ts:23]: Nit: Maybe create an `MYSQL_TYPE_ID` object to store these numerical keys. It would eliminate the need for the comments.

File: TiFBackendUtils/SQLExecutable/index.ts [hunk 5]
```diff
@@ -1,5 +1,39 @@
...
+const EnvVarsSchema = z
+  .object({
+    DATABASE_HOST: z.string(),
+    DATABASE_PASSWORD: z.string(),
+    DATABASE_USERNAME: z.string(),
+    DATABASE_NAME: z.string()
+  })
+  .passthrough()
+
+const envVars = EnvVarsSchema.parse(process.env)
```
Comment (mhayes853) [inline TiFBackendUtils/SQLExecutable/index.ts]: This should be in a separate file since we'll for sure need env vars that are not related to the DB.

File: APILambda/events/arrivals/getUpcomingEvents.ts (hunk 1 above)
Comment (seanim0920) [reply APILambda/events/arrivals/getUpcomingEvents.ts]: Added map

File: APILambda/test/database.ts (hunk 2 above)
Comment (seanim0920) [reply APILambda/test/database.ts]: Yeah very true, we weren't using DROP DATABASE here before due to Planetscale restrictions, but we no longer have those restrictions.

File: TiFBackendUtils/SQLExecutable/index.ts (hunk 5 above)
Comment (seanim0920) [reply TiFBackendUtils/SQLExecutable/index.ts]: True, I'll update after more .env vars are added

File: TiFBackendUtils/SQLExecutable/utils.ts (hunk 4 above)
Comment (seanim0920) [reply TiFBackendUtils/SQLExecutable/utils.ts:23]: I'll add in a follow up

**Review by suryatiru — APPROVED** [review]

File: APILambda/test/userFlows/events.ts (hunk 3 above)
Comment (seanim0920) [reply APILambda/test/userFlows/events.ts:44]: Played with the delay but still saw flakiness at intervals < 1000 ms.

File: APILambda/test/database.ts [hunk 6]
```diff
@@ -1,18 +1,244 @@
...
+CREATE TABLE IF NOT EXISTS userSettings (
+userId varchar(36) NOT NULL,
+isEventNotificationsEnabled tinyint(1) NOT NULL DEFAULT '1',
+isMentionsNotificationsEnabled tinyint(1) NOT NULL DEFAULT '1',
+isChatNotificationsEnabled tinyint(1) NOT NULL DEFAULT '1',
+isFriendRequestNotificationsEnabled tinyint(1) NOT NULL DEFAULT '1',
+isAnalyticsEnabled tinyint(1) NOT NULL DEFAULT '1',
+isCrashReportingEnabled tinyint(1) NOT NULL DEFAULT '1',
+updatedDateTime timestamp NULL DEFAULT current_timestamp() ON UPDATE current_timestamp(),
+PRIMARY KEY (userId),
```
Comment (seanim0920) [inline APILambda/test/database.ts:208]: use an array to store all the sql statements instead of manually adding to a separate array
